### PR TITLE
FEATURE: GitHub SSO

### PR DESCRIPTION
**Pull request by Segelzwerg**

### CODE REVIEW

File: whist_server/services/github_service.py [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+    """
+    _instance = None
+    _client_id = None
+    _client_secret = None
+    _redirect_url = None
+
+    def __new__(cls):
+        if cls._instance is None:
+            cls._instance = super(GitHubAPIService, cls).__new__(cls)
+            cls._client_id = os.getenv('CLIENT_ID')
```
Comment (iTitus) [inline whist_server/services/github_service.py:19]: Please prefix all these with `GITHUB_`

File: whist_server/services/user_db_service.py [hunk 2]
```diff
@@ -36,11 +36,23 @@ def add(cls, user: UserInDb) -> str:
...
         return UserInDb(**user)
+
+    @classmethod
+    def get_from_github(cls, username: str) -> UserInDb:
+        """
+        Similar to 'get(username)', but queries by github username instead of application username.
+        :param username: github username of the user
+        :return: the user database object
+        """
+        user = cls._users.find_one({'github_username': username})
```
Comment (iTitus) [inline whist_server/services/user_db_service.py:55]: Technically the GitHub username is not unique and we should use the ID.
Is it worth changing?

**Review by iTitus — CHANGES_REQUESTED** [review]

File: whist_server/services/user_db_service.py (hunk 2 above)
Comment (Segelzwerg) [reply whist_server/services/user_db_service.py:55]: I can have a look into it. Should create too much trouble.

File: whist_server/services/github_service.py (hunk 1 above)
Comment (Segelzwerg) [reply whist_server/services/github_service.py:19]: Will do so.

**Comment by Segelzwerg** [comment]
I'll provide more tests for the API service.

File: whist_server/services/user_db_service.py (hunk 2 above)
Comment (Segelzwerg) [reply whist_server/services/user_db_service.py:55]: > Technically the GitHub username is not unique and we should use the ID. Is it worth changing?

I had a look at the documentation. It seems they are unique: https://docs.github.com/en/site-policy/other-site-policies/github-username-policy

Comment (iTitus) [reply whist_server/services/user_db_service.py:55]: They are unique, but usernames can be changed

Comment (Segelzwerg) [reply whist_server/services/user_db_service.py:55]: Roger that, I'll check for some sort of ID and will update the identification process. 

Is the rest ok?

Comment (iTitus) [reply whist_server/services/user_db_service.py:55]: If it is too much work we can leave it as is

**Review by iTitus — CHANGES_REQUESTED** [review]
LGTM, but can you document what needs to happen to start a server instance?
What env vars need to be set? Are they required or optional? Can one disable the GitHub Login feature?

**Comment by Segelzwerg** [comment]
> LGTM, but can you document what needs to happen to start a server instance?
> 
> What env vars need to be set? Are they required or optional? Can one disable the GitHub Login feature?

Right, good point. I suggest we have this feature mandatory for now, but add a disable function later? Documentation will be added.

File: whist_server/services/user_db_service.py (hunk 2 above)
Comment (Segelzwerg) [reply whist_server/services/user_db_service.py:55]: Shouldn't be to much trouble.

Comment (Segelzwerg) [reply whist_server/services/user_db_service.py]: @iTitus I changed the behavior to identify by the ID, but still using the GitHub name during creation.

**Review by iTitus — APPROVED** [review]

**Pull request merged**